OpenRA.Mods.D2k/Traits/Sandworm.cs

```diff
@@ -0,0 +1,110 @@
+#region Copyright & License Information
+/*
+* Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+* This file is part of OpenRA, which is free software. It is made
+* available to you under the terms of the GNU General Public License
+* as published by the Free Software Foundation. For more information,
+* see COPYING.
+*/
+#endregion
+
+using System;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.D2k.Traits
+{
+	class SandwormInfo : WandersInfo
+	{
+		public const string IdleSequenceName = "idle";
+		public const int TargetRescanIntervalInTicks = 5;
+
+		public override object Create(ActorInitializer init) { return new Sandworm(init.Self, this); }
+	}
+
+	class Sandworm : Wanders, ITick, INotifyBecomingIdle
+	{
+		readonly Lazy<IMove> move;
+		readonly Lazy<RenderUnit> renderUnit;
+		readonly Lazy<AttackBase> attackTrait;
+		readonly Lazy<ReducesNoise> noiseReduction;
+
+		int targetCountdown;
+		uint currentTargetActorId;
+
+		public Sandworm(Actor self, SandwormInfo info)
+			: base(self, info)
+		{
+			move = Exts.Lazy(self.Trait<IMove>);
+			renderUnit = Exts.Lazy(self.Trait<RenderUnit>);
+			attackTrait = Exts.Lazy(self.Trait<AttackBase>);
+			noiseReduction = Exts.Lazy(() => self.World.WorldActor.Trait<ReducesNoise>());
+		}
+
+		public new void OnBecomingIdle(Actor self)
+		{
+			if (renderUnit.Value.DefaultAnimation.CurrentSequence.Name != SandwormInfo.IdleSequenceName)
+				renderUnit.Value.DefaultAnimation.ReplaceAnim("idle");
+
+			base.OnBecomingIdle(self);
+		}
+
+		public override void DoAction(Actor self, CPos targetCell)
+		{
+			Game.Debug("(Sandworm {0}) is wandering.", self.ActorID);
+			self.QueueActivity(move.Value.MoveWithinRange(Target.FromCell(self.World, targetCell, SubCell.Any), WRange.FromCells(1)));
+		}
+
+		public void Tick(Actor self)
+		{
+			if (--targetCountdown > 0)
+				return;
+
+			targetCountdown = SandwormInfo.TargetRescanIntervalInTicks;
+
+			var food = self.World.ActorsWithTrait<MakesNoise>();
```

Comment:
This will loop over the majority of actors on the map, which isn't a great approach.  It would be better to query `world.ActorsInRange` within an appropriate radius, and then filter that based on the presence of MakesNoise.